nvme.c

```diff
@@ -469,84 +559,19 @@ static int get_telemetry_log(int argc, char **argv, struct command *cmd, struct
 	else
 		err = nvme_get_host_telemetry(fd, &log);
 
-	if (err < 0)
+	if (err < 0) {
 		perror("get-telemetry-log");
-	else if (err > 0) {
+	} else if (err > 0) {
 		nvme_show_status(err);
 		fprintf(stderr, "Failed to acquire telemetry log %d!\n", err);
 		goto close_output;
 	}
 
-	switch (cfg.data_area) {
-	case 1:
-		total_size = (le16_to_cpu(log->dalb1) + 1) * NVME_LOG_TELEM_BLOCK_SIZE;
-		break;
-	case 2:
-		total_size = (le16_to_cpu(log->dalb2) + 1) * NVME_LOG_TELEM_BLOCK_SIZE;
-		break;
-	case 3:
-	default:
-		total_size = (le16_to_cpu(log->dalb3) + 1) * NVME_LOG_TELEM_BLOCK_SIZE;
-		break;
-	case 4:
-		err = nvme_identify_ctrl(fd, &ctrl);
-		if (err) {
-			perror("identify-ctrl");
-			goto close_output;
-		}
-
-		len = nvme_get_feature_length(NVME_FEAT_FID_HOST_BEHAVIOR, 0, &len);
-		if (posix_memalign(&buf, getpagesize(), len)) {
-			fprintf(stderr, "can not allocate feature payload\n");
-			errno = ENOMEM;
-			err = -1;
-			goto close_output;
-		}
-		memset(buf, 0, len);
-
-		struct nvme_get_features_args args = {
-			.args_size	= sizeof(args),
-			.fd		= fd,
-			.fid		= NVME_FEAT_FID_HOST_BEHAVIOR,
-			.nsid		= NVME_NSID_ALL,
-			.sel		= 0,
-			.cdw11		= 0,
-			.uuidx		= 0,
-			.data_len	= len,
-			.data		= buf,
-			.timeout	= NVME_DEFAULT_IOCTL_TIMEOUT,
-			.result		= &result,
-		};
-		err = nvme_get_features(&args);
-		if (err > 0) {
-			nvme_show_status(err);
-		} else if (err < 0) {
-			perror("get-feature");
-		} else {
-			if ((ctrl.lpa & 0x40)) {
-				if (((unsigned char *)buf)[1] == 1)
-					total_size = (le32_to_cpu(log->dalb4) * bs) + NVME_LOG_TELEM_BLOCK_SIZE;
-				else {
-					fprintf(stderr, "Data area 4 unsupported, Host Behavior Support ETDAS not set to 1\n");
-					errno = EINVAL;
-					err = -1;
-				}
-			} else {
-				fprintf(stderr, "Data area 4 unsupported, bit 6 of Log Page Attributes not set\n");
-				errno = EINVAL;
-				err = -1;
-			}
-		}
-		free(buf);
-		if (err)
-			goto close_output;
-		break;
-	}
-
-	err = write(output, (void *)log, total_size);
-	if (err != total_size) {
-		fprintf(stderr, "Failed to flush all data to file!\n");
-		goto close_output;
+	err = get_telemetry_log_total_size(fd, cfg.data_area, log, &total_size);
+	if (total_size) {
+		err = write(output, (void *)log, total_size);
+		if (err != total_size)
```

Comment:
Since err and total_size have two different types, it's probably better to do:
`if (err < 0 || err != total_size)`

Reply:
Fixed this also.